Production/plugins/TrainTupleProducer.cc

```diff
@@ -0,0 +1,694 @@
+            //trainTuple().genLepton_lastMotherIndex default_int_value;
+
+            const auto ref_ptr = genLepton.allParticles().data();
+
+            auto getIndex = [&](const reco_tau::gen_truth::GenParticle* particle) {
+                  int pos = -1;
+                  if(particle) {
+                      pos = static_cast<int>(particle - ref_ptr);
+                      if(pos < 0 || pos >= static_cast<int>(genLepton.allParticles().size()))
+                          throw cms::Exception("TrainTupleProducer") << "Unable to determine a gen particle index.";
+                  }
+                  return pos;
+              };
+
+              auto encodeMotherIndex = [&](const std::set<const reco_tau::gen_truth::GenParticle*>& mothers) {
+                  static constexpr int shift_scale =
+                          static_cast<Long64_t>(reco_tau::gen_truth::GenLepton::MaxNumberOfParticles);;
+
+                  if(mothers.empty()) return -1;
+                  if(mothers.size() > 6)
+                      throw cms::Exception("TrainTupleProducer") << "Gen particle with > 6 mothers.";
+                  if(mothers.size() > 1 && genLepton.allParticles().size() > static_cast<size_t>(shift_scale))
+                      throw cms::Exception("TrainTupleProducer") << "Too many gen particles per gen lepton.";
+                  int pos = 0;
+                  int shift = 1;
+                  std::set<int> mother_indices;
+                  for(auto mother : mothers)
+                      mother_indices.insert(getIndex(mother));
+                  for(int mother_idx : mother_indices) {
+                      pos = pos + shift * mother_idx;
+                      shift *= shift_scale;
+                  }
+                  return pos;
+              };
+
+              trainTuple().genLepton_lastMotherIndex.push_back(static_cast<int>(genLepton.mothers().size()) - 1);
+              for(const auto& p : genLepton.allParticles()) {
+                  trainTuple().genParticle_pdgId.push_back(p.pdgId);
+                  trainTuple().genParticle_mother.push_back(encodeMotherIndex(p.mothers));
+                  trainTuple().genParticle_charge.push_back(p.charge);
+                  trainTuple().genParticle_isFirstCopy.push_back(p.isFirstCopy);
+                  trainTuple().genParticle_isLastCopy.push_back(p.isLastCopy);
+                  trainTuple().genParticle_pt.push_back(p.p4.pt());
+                  trainTuple().genParticle_eta.push_back(p.p4.eta());
+                  trainTuple().genParticle_phi.push_back(p.p4.phi());
+                  trainTuple().genParticle_mass.push_back(p.p4.mass());
+                  trainTuple().genParticle_vtx_x.push_back(p.vertex.x());
+                  trainTuple().genParticle_vtx_y.push_back(p.vertex.y());
+                  trainTuple().genParticle_vtx_z.push_back(p.vertex.z());
+              }
+        }
+    }
+
+
+    void FillL1Paths(const std::string& var_name, bool comparison){
```

Comment:
the actual code is one line and used only once -> don't need to create a method for that